drivers/i2c/i2c_mspm0.c

```diff
@@ -0,0 +1,859 @@
+			* additional data rather than from a buffer. If the write-received
+			* function chooses to return 0 (no more data present), then 0's will
+			* be filled in
+			*/
+			if (data->target_callbacks->read_processed != NULL) {
+				uint8_t nextByte;
+
+				if (data->target_tx_valid == 0) {
+					data->target_tx_valid = data->target_callbacks->read_processed(
+						data->target_config, &nextByte);
+				}
+
+				if (data->target_tx_valid == 0) {
+					DL_I2C_transmitTargetData(config->base, nextByte);
+				} else {
+					/* In this case, no new data is desired to be filled, thus
+					* 0's are transmitted
+					*/
+					DL_I2C_transmitTargetData(config->base, 0x00);
+				}
+			} else {
+				/* read_processed function is not found. The target data will
+				 * continue to transmit to fulfill the error and not hang
+				 * the controller by stretching indefinitely */
+				DL_I2C_transmitTargetDataCheck(config->base, 0xFF);
+				data->target_tx_valid = -1;
+			}
+		}
+		break;
+	case DL_I2C_IIDX_TARGET_STOP:
+		data->state = I2C_MSPM0_IDLE;
+		k_sem_give(&data->i2c_busy_sem);
+		if (data->target_callbacks->stop) {
+			data->target_callbacks->stop(data->target_config);
+		}
+		break;
+	/* Not implemented */
+	case DL_I2C_IIDX_TARGET_RXFIFO_TRIGGER:
+	case DL_I2C_IIDX_TARGET_RXFIFO_FULL:
+	case DL_I2C_IIDX_TARGET_GENERAL_CALL:
+	case DL_I2C_IIDX_TARGET_EVENT1_DMA_DONE:
+	case DL_I2C_IIDX_TARGET_EVENT2_DMA_DONE:
+		break;
+	/* Timeout Interrupts */
+	case DL_I2C_IIDX_TIMEOUT_A:
+		data->state = I2C_MSPM0_TIMEOUT;
+
+		if(!data->is_target){
+			// Controller behavior
+			DL_I2C_disableInterrupt(config->base, TI_MSPM0_CONTROLLER_INTERRUPTS);
+			DL_I2C_clearInterruptStatus(config->base, TI_MSPM0_CONTROLLER_INTERRUPTS);
+			DL_I2C_flushControllerTXFIFO(config->base);
+
+		} else {
+			// Target Behavior
+			data->target_config->flags |= I2C_TARGET_FLAGS_ERROR_TIMEOUT;
+
+			/* the event based notify will be done using a stop condition,
+				* as there is not another known way to notify the application that a
+				* timeout has occurred, so we set a custom flag that the application
+				* must check.
+				*/
+			DL_I2C_disableInterrupt(config->base, TI_MSPM0_TARGET_INTERRUPTS);
+			DL_I2C_clearInterruptStatus(config->base, TI_MSPM0_TARGET_INTERRUPTS);
+
+			if (data->target_callbacks->stop) {
+				int stopReturn = data->target_callbacks->stop(data->target_config);
+				if (stopReturn < 0){
+					/* Allow stop to determine whether the target is reset */
+					i2c_mspm0_reset_peripheral_target(dev);
+				}
+			} else {
+				/* if no stop, just reset anyways */
+				i2c_mspm0_reset_peripheral_target(dev);
+			}
+			k_sem_give(&data->i2c_busy_sem);
+		}
+		break;
+	default:
+		break;
+	}
+}
+
+static int i2c_mspm0_init(const struct device *dev)
+{
+	const struct i2c_mspm0_config *config = dev->config;
+	struct i2c_mspm0_data *data = dev->data;
+	int ret;
+
+	k_sem_init(&data->i2c_busy_sem, 0, 1);
```

Comment:
it could be done at build time directly via Z_SEM_INITIALIZER() in the instantiation macro. (it will probably need an in-between macro though)